apps/cyberstorm-nextjs/utils/auth.ts

```diff
@@ -0,0 +1,129 @@
+import { NextResponse } from "next/server";
+import type { NextRequest } from "next/server";
+import hkdf from "@panva/hkdf";
+import { EncryptJWT, jwtDecrypt } from "jose";
+import { v4 as uuid } from "uuid";
+
+interface UserStatusType {
+  username: string | null;
+  capabilities: string[];
+  subscription: { expires: string | null };
+  rated_packages: string[];
+  teams: string[];
+}
+
+const now = () => (Date.now() / 1000) | 0;
+
+async function getDerivedEncryptionKey(secret: string | Buffer) {
+  return await hkdf(
+    "sha256",
+    secret,
+    "",
+    "Cyberstorm Generated Encryption Key",
+    32
+  );
+}
+
+export async function encode(token: { username: string; expires: number }) {
+  if (process.env.CYBERSTORM_AUTH_SECRET) {
+    const encryptionSecret = await getDerivedEncryptionKey(
+      process.env.CYBERSTORM_AUTH_SECRET
+    );
+    return await new EncryptJWT(token)
+      .setProtectedHeader({ alg: "dir", enc: "A256GCM" })
+      .setIssuedAt()
+      .setExpirationTime(Math.min(now() + 900, token.expires))
+      .setJti(uuid())
+      .encrypt(encryptionSecret);
+  } else {
+    throw new Error("CYBERSTORM_AUTH_SECRET is missing");
+  }
+}
+
+export async function decode(
+  token: string
+): Promise<Record<string, unknown> | null> {
+  if (process.env.CYBERSTORM_AUTH_SECRET) {
+    if (!token) return null;
+    const encryptionSecret = await getDerivedEncryptionKey(
+      process.env.CYBERSTORM_AUTH_SECRET
+    );
+    const { payload } = await jwtDecrypt(token, encryptionSecret, {
+      clockTolerance: 15,
+    });
+    return payload;
+  } else {
+    throw new Error("CYBERSTORM_AUTH_SECRET is missing");
+  }
+}
+
+async function fetchSubStatus(sessionid: string | undefined): Promise<{
+  username: string | null;
+  subActive: boolean;
+  expires: number;
+}> {
+  if (sessionid) {
+    const res = await fetch(
+      `https://thunderstore.io/api/experimental/current-user/`,
+      {
+        headers: {
+          authorization: `Session ${sessionid}`,
+          "Content-Type": "application/json",
+        },
+      }
+    );
+    if (res.status === 200) {
+      const currentUserStatus: UserStatusType = await res.json();
```

Comment:
You're converting `any` to `UserStatusType` on blind faith here. Some type guard or at least using nullish chaining below would be welcome. E.g. the `Date.parse` below can have odd results otherwise.

Reply:
So I made a validator, but I'm not entirely sure if validators are supposed to be made in that way. Anyways, if that's not a valid way to do it let me know 👌